File: 04/soln.go
package main

import (
	"io"
	"log"
	"os"
	"strconv"
	"strings"
)

type BoardElement struct {
	Value    int
	IsMarked bool
}

type Board struct {
	Grid       [][]*BoardElement
	IsComplete bool
}

func (b Board) Value() int {
	res := 0
	for _, row := range b.Grid {
		for _, elem := range row {
			if !elem.IsMarked {
				res += elem.Value
			}
		}
	}
	return res
}

func (b *Board) CheckRowComplete(rowIdx int) bool {
	row := b.Grid[rowIdx]
	for _, elem := range row {
		if !elem.IsMarked {
			return false
		}
	}
	b.IsComplete = true
	return true
}

func (b *Board) CheckColComplete(colIdx int) bool {
	for _, row := range b.Grid {
		if !row[colIdx].IsMarked {
			return false
		}
	}
	b.IsComplete = true
	return true
}

func (b *Board) Mark(value int) {
	for rowIdx, row := range b.Grid {
		for colIdx, elem := range row {
			if elem.Value == value {
				elem.IsMarked = true
				b.CheckColComplete(colIdx)
				b.CheckRowComplete(rowIdx)
			}
		}
	}
}

func stringsToInts(xs []string) []int {
	numbers := make([]int, 0)
	for _, s := range xs {
		number, _ := strconv.Atoi(s)
		numbers = append(numbers, number)
	}
	return numbers
}

func intsToBoardElements(xs []int) []*BoardElement {
	elements := make([]*BoardElement, 0)
	for _, num := range xs {
		element := BoardElement{Value: num, IsMarked: false}
		elements = append(elements, &element)
	}
	return elements
}

func main() {
	inputFile, err := os.Open("input.txt")
	if err != nil {
		log.Fatalln(err)
	}

	fileContents, err := io.ReadAll(inputFile)
	if err != nil {
		log.Fatalln(err)
	}

	inputLines := strings.Split(string(fileContents), "\n")
	numbersStrings := strings.Split(inputLines[0], ",")
	numbers := stringsToInts(numbersStrings)
	n_boards := (len(inputLines) - 1) / 6
	boards := make([]*Board, 0)
	for i := 0; i < n_boards; i++ {
		startIdx := 2 + i*6
		rows := make([][]*BoardElement, 0)
		for k := startIdx; k < startIdx+5; k++ {
			rowStrings := strings.Fields(inputLines[k])
			rowNums := stringsToInts(rowStrings)
			elements := intsToBoardElements(rowNums)
			rows = append(rows, elements)
		}
		board := Board{Grid: rows, IsComplete: false}
		boards = append(boards, &board)
	}

	for _, number := range numbers {
		for _, board := range boards {
			board.Mark(number)
			if board.IsComplete {
				print(board.Value() * number)
				return
			}
		}
	}

}

File: 04/soln.py
from collections.abc import Iterable
from dataclasses import dataclass


@dataclass
class BoardElement:
    value: int
    is_marked: bool


class Board:
    def __init__(self, rows: Iterable[Iterable[int]]):
        self.is_complete = False
        self.grid: list[list[BoardElement]] = []
        for row in rows:
            processed_row = [BoardElement(val, False) for val in row]
            self.grid.append(processed_row)

    @property
    def value(self) -> int:
        res = 0
        for row in self.grid:
            for elem in row:
                if not elem.is_marked:
                    res += elem.value
        return res

    def check_row_complete(self, row_idx: int) -> bool:
        row = self.grid[row_idx]
        return all(elem.is_marked for elem in row)

    def check_col_complete(self, col_idx: int) -> bool:
        return all(row[col_idx].is_marked for row in self.grid)

    def mark(self, value: int) -> None:
        for row_idx, row in enumerate(self.grid):
            for col_idx, elem in enumerate(row):
                if elem.value == value:
                    elem.is_marked = True
                    col_complete = self.check_col_complete(col_idx)
                    row_complete = self.check_row_complete(row_idx)
                    self.is_complete = col_complete or row_complete


def parse_input() -> tuple[list[int], list[Board]]:
    with open("input.txt") as f:
        inp = f.readlines()
    numbers = map(int, inp[0].split(","))
    # create boards
    boards: list[Board] = []
    for board_start in range(2, len(inp), 6):
        rows = [map(int, inp[board_start + row_num].split()) for row_num in range(5)]
        boards.append(Board(rows))
    return numbers, boards


def main():
    numbers, boards = parse_input()
    # call out numbers
    for number in numbers:
        for board in boards:
            board.mark(number)
            if board.is_complete:
                print(board.value * number)
                return


if __name__ == "__main__":
    main()
